### PR TITLE
3D Render Targets

### PR DESCRIPTION
**Pull request by MiiBond**
Enabling 3D render targets for things like voxel rendering.

I'm treating a texture with a non-zero `size.depth` as a 3D texture while a non-zero `size.layers` means a 2D texture array.

I'm using this to test:
https://playground.babylonjs.com/#XSNYAU#78

### CODE REVIEW

**Comment by bjsplat** [comment]
Please make sure to label your PR with "bug", "new feature" or "breaking change" label(s).
To prevent this PR from going to the changelog marked it with the "skip changelog" label.

**Comment by bjsplat** [comment]
Snapshot stored with reference name:
refs/pull/14897/merge

Test environment: 
https://babylonsnapshots.z22.web.core.windows.net/refs/pull/14897/merge/index.html

To test a playground add it to the URL, for example:

https://babylonsnapshots.z22.web.core.windows.net/refs/pull/14897/merge/index.html#WGZLGJ#4600

Links to test babylon tools with this snapshot:

https://playground.babylonjs.com/?snapshot=refs/pull/14897/merge
https://sandbox.babylonjs.com/?snapshot=refs/pull/14897/merge
https://gui.babylonjs.com/?snapshot=refs/pull/14897/merge
https://nme.babylonjs.com/?snapshot=refs/pull/14897/merge

To test the snapshot in the playground with a playground ID add it after the snapshot query string:

https://playground.babylonjs.com/?snapshot=refs/pull/14897/merge#BCU1XR#0

**Comment by bjsplat** [comment]
Visualization tests for WebGPU (Experimental)
Important - these might fail sporadically. This is an optional test.

https://babylonsnapshots.z22.web.core.windows.net/refs/pull/14897/merge/testResults/webgpuplaywright/index.html

**Comment by sebavan** [comment]
It should work with depth stencil as long as you also use a texture of the same dimension as a depth/stencil holder ;-) this is a limitation of DX not enforced yet in the spec. You should try to put this in place. I know it is a waste of texture mem but this is the way to enable it.

**Comment by sebavan** [comment]
all good for me to use depth

File: packages/dev/core/src/Engines/webgpuEngine.ts [hunk 1]
```diff
@@ -3220,7 +3223,7 @@ export class WebGPUEngine extends Engine {
 
         this._rttRenderPassWrapper.colorAttachmentViewDescriptor = {
             format: this._colorFormat as GPUTextureFormat,
-            dimension: WebGPUConstants.TextureViewDimension.E2d,
+            dimension: texture.is3D ? WebGPUConstants.TextureViewDimension.E3d : WebGPUConstants.TextureViewDimension.E2d,
```
Comment (MiiBond) [inline packages/dev/core/src/Engines/webgpuEngine.ts:3226]: This was hardcoded to 2D before. Does this seem like the appropriate change?

File: packages/dev/core/src/Materials/Textures/renderTargetTexture.ts [hunk 2]
```diff
@@ -779,10 +779,14 @@ export class RenderTargetTexture extends Texture implements IRenderTargetTexture
      * @returns the number of layers
      */
     public getRenderLayers(): number {
-        const layers = (<{ width: number; height: number; layers?: number }>this._size).layers;
+        const layers = (<{ width: number; height: number; depth?: number; layers?: number }>this._size).layers;
```
[the remark this answers is not in the document]
Comment (MiiBond) [reply packages/dev/core/src/Materials/Textures/renderTargetTexture.ts:782]: Done

File: packages/dev/core/src/Engines/WebGPU/webgpuHardwareTexture.ts [hunk 3]
```diff
@@ -88,6 +88,7 @@ export class WebGPUHardwareTexture implements HardwareTextureWrapper {
             arrayLayerCount = 6 * (depth || 1);
         } else if (is3D) {
             viewDimension = WebGPUConstants.TextureViewDimension.E3d;
+            arrayLayerCount = depth;
```
[the remark this answers is not in the document]
Comment (MiiBond) [reply packages/dev/core/src/Engines/WebGPU/webgpuHardwareTexture.ts]: Done. Thank you!

File: packages/dev/core/src/Engines/WebGPU/webgpuTextureManager.ts [hunk 4]
```diff
@@ -1069,7 +1069,7 @@ export class WebGPUTextureManager {
 
             gpuTextureWrapper.set(gpuTexture);
 
-            const arrayLayerCount = texture.is3D ? 1 : layerCount;
+            const arrayLayerCount = layerCount;
```
[the remark this answers is not in the document]
Comment (MiiBond) [reply packages/dev/core/src/Engines/WebGPU/webgpuTextureManager.ts]: Done

**Comment by bjsplat** [comment]
Visualization tests for WebGPU (Experimental)
Important - these might fail sporadically. This is an optional test.

https://babylonsnapshots.z22.web.core.windows.net/refs/pull/14897/merge/testResults/webgpuplaywright/index.html

**Comment by sebavan** [comment]
I like where it is going, I leave the final approval to @Popov72.

I wonder if a visual test could be added ?

File: packages/dev/core/src/Engines/webgpuEngine.ts (hunk 1 above)
Comment (RaananW) [reply packages/dev/core/src/Engines/webgpuEngine.ts:3226]: pinging @Popov72

**Comment by MiiBond** [comment]
I've added a visual test but, for WebGPU, it won't work unless the browser has the unsafe-webgpu flag enabled so I'm not sure how to generate a reference screenshot.

**Comment by Popov72** [comment]
> 've added a visual test but, for WebGPU, it won't work unless the browser has the unsafe-webgpu flag enabled so I'm not sure how to generate a reference screenshot.

That's ok, we will disable this test for WebGPU if it does not work on the CI.

I had to make some small changes for WebGPU, here they are (it was easier for me to do them locally than to comment on the PR):

https://github.com/Popov72/Babylon.js/compare/a10bfc77d91631510845e2a877273706b6e9242e...06e18e8919327552946a3c9b7219db33a2191fb1

With these changes, this PG will work in WebGPU: http://localhost:1338/?webgpu#XSNYAU#91

It won't work in WebGL, though, as it seems you can't render to different types of textures in a MRT in WebGL (the PG renders to two slices of a 3D texture, one face of a cube texture and in a 2D texture).

**Comment by RaananW** [comment]
Merging from master will help with resolving this conflict, and then we can see if it is running correctly on the CI for WebGPU

**Comment by MiiBond** [comment]
Thank you. I've merged from master and incorporated @Popov72's fixes.

File: packages/dev/core/src/Engines/thinEngine.ts [hunk 5]
```diff
@@ -1963,8 +1963,17 @@ export class ThinEngine {
 
         const depthStencilTexture = rtWrapper._depthStencilTexture;
         if (depthStencilTexture) {
+            if (rtWrapper.is3D) {
+                if (
+                    rtWrapper.texture!.width != depthStencilTexture.width ||
```
Comment (Popov72) [inline packages/dev/core/src/Engines/thinEngine.ts:1968]: nit: `!==` instead of `!=`

File: packages/tools/tests/test/visualization/config.json [hunk 6]
```diff
@@ -2036,6 +2042,12 @@
             "title": "CesiumMan from Khronos Sample Assets",
             "playgroundId": "#4GWX8M",
             "referenceImage": "CesiumMan.png"
+        },
+        {
+            "title": "Render to 3D RT",
+            "renderCount": 5,
```
Comment (Popov72) [inline packages/tools/tests/test/visualization/config.json:2048]: This test is duplicated in the file.

**Review by Popov72 — CHANGES_REQUESTED** [review]

**Review by Popov72 — APPROVED** [review]

**Comment by bjsplat** [comment]
Visualization tests for WebGL 1 have failed. If some tests failed because the snapshots do not match, the report can be found at

https://babylonsnapshots.z22.web.core.windows.net/refs/pull/14897/merge/testResults/webgl1/index.html

If tests were successful afterwards, this report might not be available anymore.

**Pull request merged**